crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1392,48 +1477,137 @@ impl<'db> Type<'db> {
+                &CallArguments::positional([instance.unwrap_or(Type::none(db)), owner]),
+            )
+            .return_type(db)
+    }
+
+    /// Access an attribute of this type, potentially invoking the descriptor protocol.
+    /// Corresponds to `getattr(<object of type 'self'>, name)`.
+    ///
+    /// See also: [`Type::static_member`]
+    #[must_use]
+    pub(crate) fn member(&self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        if name == "__class__" {
+            return Symbol::bound(self.to_meta_type(db));
+        }
+
+        match self {
+            Type::FunctionLiteral(function) if name == "__get__" => Symbol::bound(Type::Callable(
+                CallableType::MethodWrapperDunderGet(*function),
+            )),
+
+            Type::Callable(CallableType::BoundMethod(bound_method)) => match name {
+                "__self__" => Symbol::bound(bound_method.self_instance(db)),
+                "__func__" => Symbol::bound(Type::FunctionLiteral(bound_method.function(db))),
+                _ => KnownClass::MethodType.to_instance(db).member(db, name),
+            },
+            Type::Callable(CallableType::MethodWrapperDunderGet(_)) => {
+                KnownClass::MethodWrapperType
+                    .to_instance(db)
+                    .member(db, name)
+            }
+            Type::Callable(CallableType::WrapperDescriptorDunderGet) => {
+                KnownClass::WrapperDescriptorType
+                    .to_instance(db)
+                    .member(db, name)
+            }
+
+            Type::Instance(..)
+            | Type::BooleanLiteral(..)
+            | Type::IntLiteral(..)
+            | Type::StringLiteral(..)
+            | Type::BytesLiteral(..)
+            | Type::LiteralString
+            | Type::SliceLiteral(..)
+            | Type::Tuple(..)
+            | Type::KnownInstance(..)
+            | Type::FunctionLiteral(..) => {
+                let member = self.static_member(db, name);
+
+                let instance = Some(*self);
+                let owner = self.to_meta_type(db);
+
+                member.map_type(|ty| ty.try_call_dunder_get(db, instance, owner).unwrap_or(ty))
+            }
+            Type::ClassLiteral(..) => {
+                let member = self.static_member(db, name);
+
+                let instance = None;
+                let owner = self.to_meta_type(db);
+
+                member.map_type(|ty| ty.try_call_dunder_get(db, instance, owner).unwrap_or(ty))
+            }
+            Type::Union(union) => union.map_with_boundness(db, |elem| elem.member(db, name)),
+            Type::Intersection(..) => Symbol::todo("Attribute access on `Intersection` types"),
+
+            Type::Dynamic(..)
+            | Type::Never
+            | Type::AlwaysFalsy
+            | Type::AlwaysTruthy
+            | Type::ModuleLiteral(..) => self.static_member(db, name),
+
+            // TODO: We fall back to static member for `type[…]` for now, but this could be improved.
```

Comment:
Is there a reason we don't handle this the same as `Type::ClassLiteral`? Is that incorrect for a reason I'm missing, or does it result in some problems?

Reply:
> Is there a reason we don't handle this the same as `Type::ClassLiteral`?

Not really, except that I thought you mentioned something regarding `SubtypeOf` being slightly more complex (https://github.com/astral-sh/ruff/issues/15948?), and that it would be okay to leave it as a TODO. But I was now "forced" to move this pattern to the `ClassLiteral` branch due to some test failures, so your reasoning seems correct :+1:

Reply:
Oh, calling `SubclassOf` is more complex because `__init__` and `__new__` overrides aren't required to obey Liskov, so calling a `SubclassOf` type is inherently unsound. I don't think this applies to member access or descriptor handling, though. (Well, potentially it could apply to member access of `__init__` or `__new__` method on a `SubclassOf` type, since this would also be unsound. But that would be handled as a special case, doesn't need to affect anything in this PR.)